### ISSUE TITLE
`not-an-iterable` false positive with class methods with @typing.overload decorator

### ISSUE THREAD

**Issue by Parnassius** [PyCQA/pylint#8314]
### Bug description

```py
# pylint: disable=missing-module-docstring,missing-class-docstring,missing-function-docstring,too-few-public-methods,invalid-name,unused-argument

from typing import overload

class A:
    @overload
    def func(self, arg: int) -> list[int]:
        ...

    @overload
    def func(self, arg: str) -> list[int]:
        ...

    def func(self, arg: str | int) -> list[int]:
        return [1, 2, 3]

a = A()

for x in a.func(1):
    pass
```

The typing overloads don't really make sense here, but they are required to reproduce the issue.

### Command used

```shell
pylint class_overload.py
```

### Pylint output

```shell
************* Module class_overload
class_overload.py:19:9: E1133: Non-iterable value a.func(1) is used in an iterating context (not-an-iterable)

-----------------------------------
Your code has been rated at 5.45/10
```

### Expected behavior

No error.

The issue only appears with an `@overload`ed class method, the following three similar situations work correctly:

Class method with no overloads:
```py
# pylint: disable=missing-module-docstring,missing-class-docstring,missing-function-docstring,too-few-public-methods,invalid-name,unused-argument

class A:
    def func(self, arg: str | int) -> list[int]:
        return [1, 2, 3]

a = A()

for x in a.func(1):
    pass
```

Function with overloads:
```py
# pylint: disable=missing-module-docstring,missing-class-docstring,missing-function-docstring,too-few-public-methods,invalid-name,unused-argument

from typing import overload

@overload
def func(arg: int) -> list[int]:
    ...

@overload
def func(arg: str) -> list[int]:
    ...

def func(arg: str | int) -> list[int]:
    return [1, 2, 3]

for x in func(1):
    pass
```

Function with no overloads:
```py
# pylint: disable=missing-module-docstring,missing-class-docstring,missing-function-docstring,too-few-public-methods,invalid-name,unused-argument

def func(arg: str | int) -> list[int]:
    return [1, 2, 3]

for x in func(1):
    pass
```

### Pylint version

```shell
pylint 2.16.2
astroid 2.14.2
Python 3.10.9 (main, Dec 08 2022, 14:49:06) [GCC]

The issue is also present on the latest commit (490c40525c286483e8b7405870a8d2f0ae627de4)
```

### OS / Environment

OpenSUSE Tumbleweed

**Comment by jacobtylerwalls** [comment]
Duplicate of PyCQA/astroid#1015
